backend/cfg/cfg_cse.ml

```diff
@@ -0,0 +1,250 @@
+
+class cse_generic =
+  object (self)
+    method class_of_operation : Cfg.operation -> op_class =
+      function
+      | Move | Spill | Reload -> assert false (* treated specially *)
+      | Const_int _ | Const_float _ | Const_symbol _ | Const_vec128 _ -> Op_pure
+      | Opaque -> assert false (* treated specially *)
+      | Stackoffset _ -> Op_other
+      | Load { mutability; is_atomic; memory_chunk = _; addressing_mode = _ } ->
+        (* #12173: disable CSE for atomic loads. *)
+        if is_atomic
+        then Op_other
+        else
+          Op_load
+            (match mutability with
+            | Mutable -> Mutable
+            | Immutable -> Immutable)
+      | Store (_, _, asg) -> Op_store asg
+      | Alloc _ | Poll -> assert false (* treated specially *)
+      | Intop _ -> Op_pure
+      | Intop_imm (_, _) -> Op_pure
+      | Intop_atomic _ -> Op_store true
+      | Compf _ | Csel _ | Negf | Absf | Addf | Subf | Mulf | Divf
+      | Scalarcast _ | Floatofint | Intoffloat | Valueofint | Intofvalue
+      | Vectorcast _ ->
+        Op_pure
+      | Specific _ -> Op_other
+      | Name_for_debugger _ -> Op_other
+      | Probe_is_enabled _ -> Op_other
+      | Begin_region | End_region -> Op_other
+      | Dls_get -> Op_load Mutable
+
+    method is_cheap_operation : Cfg.operation -> bool =
+      function Const_int _ -> true | _ -> false
+    [@@warning "-4"]
+
+    method private kill_loads (n : numbering) : numbering =
+      remove_mutable_load_numbering n
+
+    method private cse_instruction
+        : State.t ->
+          numbering ->
+          Cfg.basic Cfg.instruction DLL.cell ->
+          numbering =
+      fun state n cell ->
+        let i = DLL.value cell in
+        match i.desc with
+        | Reloadretaddr | Pushtrap _ | Poptrap | Prologue -> n
+        | Op (Move | Spill | Reload) ->
+          let n1 = set_move n i.arg.(0) i.res.(0) in
+          n1
+        | Op Opaque -> empty_numbering
+        | Op (Alloc _) | Op Poll ->
+          let n1 = kill_addr_regs (self#kill_loads n) in
+          let n2 = set_unknown_regs n1 i.res in
+          n2
+        | Op op -> (
+          match self#class_of_operation op with
+          | (Op_pure | Op_load _) as op_class -> (
+            let n1, varg = valnum_regs n i.arg in
+            let n2 = set_unknown_regs n1 (Proc.destroyed_at_basic i.desc) in
+            match find_equation op_class n1 (op, varg) with
+            | Some vres -> (
+              (* This operation was computed earlier. *)
+              (* Are there registers that hold the results computed earlier? *)
+              match find_regs_containing n1 vres with
+              | Some res when not (self#is_cheap_operation op) ->
+                (* We can replace res <- op args with r <- move res, provided
+                   res are stable (non-volatile) registers. If the operation is
+                   very cheap to compute, e.g. an integer constant, don't
+                   bother. *)
```

Comment:
I don't understand this comment (it's originally from CSEgen), the reference to  `res` twice doesn't make sense and `r` is not around. The code for this case is fine.

Reply:
I would probably write something
along the lines of "We can replace
the operation with a move, provided
the registers are stable (non-volatile).
If the operation is very cheap to
compute, e.g. an integer constant,
don't bother."

Reply:
Sounds good.